apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerReporter.java

```diff
@@ -155,83 +168,79 @@ public long getReported() {
-        }
-        final long cursor = disruptor.getCursor();
-        return new Future<Void>() {
-            private volatile boolean cancelled = false;
-
-            @Override
-            public boolean cancel(boolean mayInterruptIfRunning) {
-                if (isDone()) {
-                    return false;
-                }
-                disruptor.get(cursor).resetState();
-                // the volatile write also ensures visibility of the resetState() in other threads
-                cancelled = true;
-                return true;
-            }
-
-            @Override
-            public boolean isCancelled() {
-                return cancelled;
-            }
+    public void scheduleWakeupEvent() {
+        disruptor.getRingBuffer().tryPublishEvent(WAKEUP_EVENT_TRANSLATOR);
+    }
 
-            @Override
-            public boolean isDone() {
-                return isEventProcessed(cursor);
-            }
+    @Override
+    public boolean hardFlush(long timeout, TimeUnit unit) {
+        return publishAndWaitForEvent(timeout, unit, END_REQUEST_EVENT_TRANSLATOR);
+    }
 
-            @Override
-            public Void get() throws InterruptedException {
-                while (!isDone()) {
-                    Thread.sleep(1);
-                }
-                return null;
-            }
+    @Override
+    public boolean softFlush(long timeout, TimeUnit unit) {
+        return publishAndWaitForEvent(timeout, unit, SYNC_FLUSH_EVENT_TRANSLATOR);
+    }
 
-            /*
-             * This might not a very elegant or efficient implementation but it is only intended to be used in tests anyway
-             */
-            @Override
-            public Void get(long timeout, TimeUnit unit) throws InterruptedException, TimeoutException {
-                for (; timeout > 0 && !isDone(); timeout--) {
-                    Thread.sleep(1);
+    private boolean publishAndWaitForEvent(long timeout, TimeUnit unit, EventTranslatorOneArg<ReportingEvent, Thread> eventTranslator) {
+        if (!reportingEventHandler.isHealthy()) {
+            return false;
+        }
+        ReportingEventHandler reportingEventHandler = this.reportingEventHandler;
+        long startNs = System.nanoTime();
+        long thresholdNs;
+        if (timeout < 0) {
+            thresholdNs = Long.MAX_VALUE;
+        } else {
+            thresholdNs = unit.toNanos(timeout) + startNs;
+        }
+        do {
+            try {
+                long sequence = disruptor.getRingBuffer().tryNext();
+                try {
+                    eventTranslator.translateTo(disruptor.get(sequence), sequence, Thread.currentThread());
+                } finally {
+                    disruptor.getRingBuffer().publish(sequence);
                 }
-                if (!isDone()) {
-                    throw new TimeoutException();
+                return waitForEventProcessed(sequence, thresholdNs);
```

Comment:
This creates a race condition- the event has already been published and potentially processed by the disruptor handler thread (including attempting to `unpark`) before this thread gets the chance to `park`.

Reply:
I don't think so. Note that `waitForEventProcessed` checks if the sequence has already been processed. It's always required to check the condition after park returns as a thread may be unparked for random reasons.

Reply:
Right! I missed the check in the for-loop stop condition, only saw the one outside the for-loop 👍